data/mods/gen1/scripts.ts

```diff
@@ -68,531 +58,742 @@ export const Scripts: ModdedBattleScriptsData = {
+				}
+
+				// Now we check for the move's critical hit ratio.
+				if (move.critRatio === 1) {
+					// Normal hit ratio, we divide the crit chance by 2 and floor the result again.
+					critChance = Math.floor(critChance / 2);
+				} else if (move.critRatio === 2) {
+					// High crit ratio, we multiply the result so far by 4 and cap it at 255.
+					critChance = this.battle.clampIntRange(critChance * 4, 1, 255);
+				}
+
+				// Last, we check deppending on ratio if the move critical hits or not.
+				// We compare our critical hit chance against a random number between 0 and 255.
+				// If the random number is lower, we get a critical hit. This means there is always a 1/255 chance of not hitting critically.
+				if (critChance > 0) {
+					isCrit = this.battle.randomChance(critChance, 256);
 				}
 			}
-		}
-		if (move.selfSwitch && pokemon.hp) {
-			pokemon.switchFlag = move.selfSwitch;
-		}
+			if (isCrit) target.getMoveHitData(move).crit = true;
+
+			// Happens after crit calculation.
+			if (basePower) {
+				basePower = this.battle.runEvent('BasePower', pokemon, target, move, basePower);
+				if (basePower && move.basePowerModifier) {
+					basePower *= move.basePowerModifier;
+				}
+			}
+			if (!basePower) return 0;
+			basePower = this.battle.clampIntRange(basePower, 1);
+
+			// We now check attacker's and defender's stats.
+			let level = pokemon.level;
+			let attacker = pokemon;
+			const defender = target;
+			if (move.useTargetOffensive) attacker = target;
+			const atkType: StatNameExceptHP = (move.category === 'Physical') ? 'atk' : 'spa';
+			const defType: StatNameExceptHP = (move.defensiveCategory === 'Physical') ? 'def' : 'spd';
+			let attack = attacker.getStat(move.useSourceDefensiveAsOffensive ? defType : atkType);
+			let defense = defender.getStat(defType);
+			// In gen 1, screen effect is applied here.
+			if ((defType === 'def' && defender.volatiles['reflect']) || (defType === 'spd' && defender.volatiles['lightscreen'])) {
+				this.battle.debug('Screen doubling (Sp)Def');
+				defense *= 2;
+				defense = this.battle.clampIntRange(defense, 1, 1998);
+			}
+
+			// In the event of a critical hit, the offense and defense changes are ignored.
+			// This includes both boosts and screens.
+			// Also, level is doubled in damage calculation.
+			if (isCrit) {
+				move.ignoreOffensive = true;
+				move.ignoreDefensive = true;
+				level *= 2;
+				if (!suppressMessages) this.battle.add('-crit', target);
+			}
+			if (move.ignoreOffensive) {
+				this.battle.debug('Negating (sp)atk boost/penalty.');
+				attack = attacker.getStat(atkType, true);
+			}
+			if (move.ignoreDefensive) {
+				this.battle.debug('Negating (sp)def boost/penalty.');
+				// No screens
+				defense = target.getStat(defType, true);
+			}
+
+			// When either attack or defense are higher than 256, they are both divided by 4 and moded by 256.
+			// This is what cuases the roll over bugs.
```

Comment:
```suggestion
			// This is what causes the roll over bugs.
```
Might as well fix this typo while you're changing the file anyway.

Reply:
I mean, "moded" is like "plused" – it's bad grammar but it might be shorter/easier for people to understand than "and then take the remainder modulo 256".

Reply:
At least in some dictionaries, the verb "to plus" exists, so "plused" isn't bad grammar. However the existing verb "to mod" used as an abbreviation of either "moderate" or "modify" has the simple past tense "modded" rather than "moded" (which would have been the simple past tense of "to mode"). While you could have used "reduced mod 256", I like the new comment anyway.